### PR TITLE
Move error handling out of observable and retry mongo

### PR DESCRIPTION
**Pull request by philwills**
This is yet another attempt to fix CD.

### CODE REVIEW

**Comment by philwills** [comment]
cc @guardian/deploy-infrastructure

**Comment by cb372** [comment]
Fingers crossed! 

Have you tried deploying this branch? Did it help?

**Comment by philwills** [comment]
No, as I've been out, I haven't yet given it a whirl. Just rebased, so I can, but am in meetings most of today.

File: riff-raff/app/ci/ContinuousDeployment.scala [hunk 1]
```diff
@@ -88,5 +75,12 @@ object ContinuousDeployment extends LifecycleWithoutApp with Logging {
       config.findMatchOnSuccessfulBuild(build).map(build => config -> build)
     }
   }
+
+  def retryUpTo[T](maxAttempts: Int)(thunk: () => T): Try[T] = {
+    val thunkStream = Stream.continually(Try(thunk()))
+      .take(maxAttempts)
+
+    thunkStream.find(_.isSuccess).getOrElse(thunkStream.head)
```
Comment (adamnfish) [inline riff-raff/app/ci/ContinuousDeployment.scala]: I think this should noisily log helpful info if it exceeds its retry count

Comment (adamnfish) [inline riff-raff/app/ci/ContinuousDeployment.scala]: Oh don't worry, I see you're doing it from the caller.

File: riff-raff/app/ci/ContinuousDeployment.scala [hunk 2]
```diff
@@ -8,47 +8,34 @@ import persistence.Persistence.store.getContinuousDeploymentList
...
-        buildsToDeploy(buildCandidates(CIBuild.newBuilds))
-      }
-    }
+    val builds = buildCandidates(CIBuild.newBuilds)
 
-    sub = Some(builds.subscribe ({ x =>
-      runDeploy(getDeployParams(x))
-    }, e => log.error("Problem looking for new CD candidates", e)))
+    val cdConfigs = retryUpTo(5)(getContinuousDeploymentList).getOrElse{
+      log.error("Failed to retrieve CD configs")
```
Comment (adamnfish) [inline riff-raff/app/ci/ContinuousDeployment.scala]: Probably useful to log the actual error if possible.

File: riff-raff/app/ci/ContinuousDeployment.scala [hunk 3]
```diff
@@ -88,5 +75,12 @@ object ContinuousDeployment extends LifecycleWithoutApp with Logging {
       config.findMatchOnSuccessfulBuild(build).map(build => config -> build)
     }
   }
+
+  def retryUpTo[T](maxAttempts: Int)(thunk: () => T): Try[T] = {
+    val thunkStream = Stream.continually(Try(thunk()))
```
Comment (adamnfish) [inline riff-raff/app/ci/ContinuousDeployment.scala]: Nice use of `Stream`

**Comment by philwills** [comment]
This seems to be working, so I'm going to merge.

**Pull request merged**